### ISSUE TITLE
Missing image is raising 500 on s3

### ISSUE THREAD

**Issue by shredding** [wagtail/wagtail#2166]
In `wagtailimages/models.py` catching the OSError is not enough for non-existing images for S3 backed CDNs.

Namely the `get_file_size` method digs into s3boto storage and raises an Attribute Error here: https://github.com/jschneier/django-storages/blob/master/storages/backends/s3boto.py#L473

It's arguably an inexpressive error to raise here, however, is catching that error an option? 

This _Work on my machine_™ (only difference is the added catching of the AttributeError):

``` python
    def get_file_size(self):
        if self.file_size is None:
            try:
                self.file_size = self.file.size
            except (OSError, AttributeError):
                # File doesn't exist
                return

            self.save(update_fields=['file_size'])

        return self.file_size
```

**Comment by plumdog** [comment]
The exception that gets thrown depends on the `Storage` subclass you happen to be using. So if you're using `boto3`, you'll get a `ClientError` (from `botocore.exceptions`). If you've written your own custom storage, you could throw any old exception of your choice.

My hunch is that this is a failing in Django's storages, in that it isn't clear what something like `size(name)` should do if `name` doesn't exist, .

So I think this code kinda has to just become:
```python
try:
    size = self.file.size
except Exception:
    # Catch any error from the storage, we assume that means the file doesn't exist
    return
# Separate this out so if - for some reason - setting the attribute errors, that won't be silently ignored
self.file_size = size
self.save(update_fields['file_size'])

**Comment by gasman** [comment]
Thanks for the PR, @plumdog.

I have to wonder: wouldn't it make more sense to go the other way, and not catch the exception at all? If the file is missing for some reason, then that's an error condition, and should be reported as such.

Alternatively, if we want to take a more developer-friendly stance (missing file _is_ an error, but it's a mundane one that will frequently occur when copying databases between production/staging/developer environments, and it doesn't justify breaking the image-edit view or spamming the developer with error emails), we should probably throw a predictable exception from `get_file_size` and have the view code catch it. As it stands, `get_file_size` is returning `None` to indicate that the file size couldn't be determined, which feels like a misuse of return values.

**Comment by plumdog** [comment]
@gasman Hi!

Yeah, a predictable exception would probably be the best of all worlds, but is a change to the API.

It looks like `SourceImageIOError` already exists for this purpose.

I'll push a commit for this. If it looks reasonable I'll write some tests.

**Comment by chosak** [comment]
Thanks @plumdog! Your PR #4468 fixes this and was merged (with some additional unit tests) in 24712b4.
